### PR TITLE
fix(autocomplete): update template when changing autocomplete in trigger

### LINKED ISSUE #13812: Changing the autocomplete on a trigger does not change the overlay (options)
#### Bug, feature request, or proposal:
Changing the autocomplete on a trigger does not change the options in the overlay.

#### What is the expected behavior?
When changing the autocomplete (`dtAutocomplete` input) on a trigger, the overlay should display the new options.

#### What is the current behavior?
When changing the autocomplete (`dtAutocomplete` input) on a trigger, the overlay options
are still the initial ones.

#### What are the steps to reproduce?
https://stackblitz.com/edit/angular-7djgzt?file=app%2Fautocomplete-simple-example.html

#### Which versions of Angular, Material, OS, TypeScript, browsers are affected?
Current

### PR DESCRIPTION
**Pull request by thomaspink**
fixes #13812

### CODE REVIEW

**Comment by thomaspink** [comment]
I signed it!

File: src/lib/autocomplete/autocomplete-trigger.ts [hunk 1]
```diff
@@ -569,8 +569,15 @@ export class MatAutocompleteTrigger implements ControlValueAccessor, OnDestroy {
       throw getMatAutocompleteMissingPanelError();
     }
 
-    if (!this._overlayRef) {
+    if (!this._portal || this._portal.templateRef !== this.autocomplete.template) {
```
Comment (crisbeto) [inline src/lib/autocomplete/autocomplete-trigger.ts]: A cleaner approach to this would be to move the `_portal` to `MatAutocomplete` and have the trigger just call `OverlayRef.attach`.

File: src/lib/autocomplete/autocomplete.spec.ts [hunk 2]
```diff
@@ -2607,3 +2639,19 @@ class AutocompleteWithNativeAutocompleteAttribute {
...
+      <mat-option [value]="0">First</mat-option>
+    </mat-autocomplete>
+
+    <mat-autocomplete #auto1="matAutocomplete">
+      <mat-option [value]="1">Second</mat-option>
+    </mat-autocomplete>
+  `,
+})
+class DynamicallyChangingAutocomplete {
+  selected = 0;
```
Comment (crisbeto) [inline src/lib/autocomplete/autocomplete.spec.ts]: This seems like a weird way to toggle between the panels. What you can do instead is to get a reference to both of them through a `viewChild` and assign them directly as properties.

File: src/lib/autocomplete/autocomplete-trigger.ts (hunk 1 above)
Comment (crisbeto) [reply src/lib/autocomplete/autocomplete-trigger.ts]: I would've preferred to move it for the menu as well, but it has a different set of requirements, because the consumer can pass in any object that implements the `MatMenuPanel` interface.

File: src/lib/autocomplete/autocomplete.spec.ts (hunk 2 above)
Comment (thomaspink) [reply src/lib/autocomplete/autocomplete.spec.ts]: fixed

File: src/lib/autocomplete/autocomplete-trigger.ts (hunk 1 above)
Comment (thomaspink) [reply src/lib/autocomplete/autocomplete-trigger.ts]: I moved the portal to the `mat-autocomplete`. But we still have to detach first in the trigger before attaching a new portal, otherwise it will throw the "already attached" error.
I am also un- and resubscribing to closing actions now when attaching a new portal, so we get the correct option changes events.

File: src/lib/autocomplete/autocomplete-trigger.ts [hunk 3]
```diff
@@ -596,9 +595,19 @@ export class MatAutocompleteTrigger implements ControlValueAccessor, OnDestroy {
       this._overlayRef.updateSize({width: this._getPanelWidth()});
     }
 
-    if (this._overlayRef && !this._overlayRef.hasAttached()) {
-      this._overlayRef.attach(this._portal);
+    if (this._currentPortalAttached !== this.autocomplete._portal) {
```
Comment (crisbeto) [inline src/lib/autocomplete/autocomplete-trigger.ts]: I'm not sure why you'd need to do this. You can just call `detach` and then `attach` again to set the new portal.

Comment (thomaspink) [reply src/lib/autocomplete/autocomplete-trigger.ts]: I see your point there, but i am currently working on an issue (appears with and without this check) where a new panel will be created, even if the template has not changed. This especially happens in the tests, thats why a lot of them are failing.

For example take a look at this test: [autocomplete.spec.ts#L267](https://github.com/angular/material2/blob/master/src/lib/autocomplete/autocomplete.spec.ts#L267). 
It is failing because we are creating a new panel on `focusin` and later on `typeInElement` (`_attachOverlay` gets called in multiple times).
So i need to find a way to check if we should detach/reatach (when a new `mat-autocomplete` has been set or not.
Maybe you have a good solution/idea for this one? Detach only if the template has changed?

Comment (crisbeto) [reply src/lib/autocomplete/autocomplete-trigger.ts]: You don't have to do it in `_attachPortal` though. Keeping `_attachPortal` as it is in master, aside from taking the portal from the panel, and calling `detach` when the `autocomplete` input changes should do it.

Comment (thomaspink) [reply src/lib/autocomplete/autocomplete-trigger.ts]: Changed it to a detach in the autocomplete input. Thanks for your help

File: src/lib/autocomplete/autocomplete-trigger.ts [hunk 4]
```diff
@@ -166,7 +164,15 @@ export class MatAutocompleteTrigger implements ControlValueAccessor, OnDestroy {
   _onTouched = () => {};
 
   /** The autocomplete panel to be attached to this trigger. */
-  @Input('matAutocomplete') autocomplete: MatAutocomplete;
+  @Input('matAutocomplete')
+  get autocomplete(): MatAutocomplete { return this._autocomplete; }
+  set autocomplete(value: MatAutocomplete) {
+    this._autocomplete = value;
+    if (this._overlayRef) {
+      this._overlayRef.detach();
```
Comment (thomaspink) [inline src/lib/autocomplete/autocomplete-trigger.ts]: @crisbeto  Not sure if it would be better to do call `closePanel()` here instead of just `detach` so the values of `_overlayAttached` and `autocomplete._isOpen` are getting updated as well.

File: src/lib/autocomplete/autocomplete.spec.ts [hunk 5]
```diff
@@ -2221,6 +2221,38 @@ describe('MatAutocomplete', () => {
...
+
+    // close overlay
+    dispatchFakeEvent(document, 'click');
+    fixture.detectChanges();
+
+    // Switch to second autocomplete
+    fixture.componentInstance.trigger.autocomplete = fixture.componentInstance.autoTow;
+    fixture.detectChanges();
+
+    // reopen agian
```
Comment (crisbeto) [inline src/lib/autocomplete/autocomplete.spec.ts]: agian -> again. Alternately these comments can be removed since they don't contribute much.

**Comment by mmalerba** [comment]
This was accidentally merged despite breaking some applications inside Google, so I had to revert it. 

@thomaspink feel free to open a new PR and we can try to resolve the breakages before merging again
